Add scaffolder option to display object items in separate rows on review page

stephenglass opened the pull request:
## Hey, I just made a Pull Request!

### Use Case

We have a use case where we are creating dynamic UI for a template using a custom field extension. In this case the field would return an object with many items. The current behavior by Backstage is to display an object in a single row on the review page. 

We would instead like to have each item in the object display on separate rows on the review page similarly to the default behavior for primitive types. Another example is using objects in template to organize your fields and your template design, but you still want each object value to display in a separate row. There are likely many users that would prefer (nested) objects to display in this way. 

### Solution

- Display each item in an object on a separate row by default
- Option is configurable at every level for deeply nested objects
- Supports deeply nested objects
- Apply any review state logic (like masking and hiding fields) to children in objects which previously was not applied

### Example
Template options:

```yaml
  parameters:
    - title: Fill in some steps
      properties:
        configuration:
          title: Configuration
          type: object
          properties:
            firstName:
              title: First Name
              type: string
            lastName:
              title: LastName
              type: string
            employeeData:
              title: Employee Data
              type: object
              properties:
                employeeNumber:
                  title: Employee Number
                  type: string
                isContractor:
                  title: Contractor
                  type: boolean
            personalData:
              title: Personal Data
              type: object
              ui:backstage:
                review:
                  # don't explode the personal data object
                  explode: false
              properties:
                creditCard:
                  title: Credit Card
                  type: string
                accountNumber:
                  title: Account Number
                  type: string
        regularField:
          title: Regular Field
          type: string
```

### Form UI (no changes here)
![image](https://github.com/backstage/backstage/assets/31711200/ad49975f-2adc-4ba2-a15b-fae3b6e7171f)



### Updated Review Page UI 

- `firstName`, `lastName`, `employeenumber`, `isContractor` display on separate rows because nested objects are exploded by default
- `creditCard`, `accountNumber` display on a single row (current behavior) because `ui:backstage.review.explode` was explicitly set to `false`

![image](https://github.com/backstage/backstage/assets/31711200/f49e97e9-0aca-4ca9-8e73-bfed47d85e85)


Appreciate any feedback as always!


#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

acierto commented on plugins/scaffolder-react/src/next/components/ReviewState/ReviewState.tsx:
```diff
@@ -28,14 +28,47 @@ export type ReviewStateProps = {
   formState: JsonObject;
 };
 
+function flattenObject(
```
It worths to move this logic to some separate file, `util.ts` or `flatten.ts` and cover it separately with a unit test.
The less logic in UI component is better, easier to comprehend what is going on a representation side.

stephenglass replied:
Sure. Any suggestions where a good place would be to place this file?

acierto replied:
I would keep this file in the same folder as it still belongs to this component only

stephenglass replied:
Done!

acierto commented:
Hey @stephenglass, this case might be useful indeed.
UX wise it might be better to display table in table. 
Did you test that booleans and other types display consistently with the current behaviour (like a check mark icon)?

stephenglass commented:
> Hey @stephenglass, this case might be useful indeed.
> 
> UX wise it might be better to display table in table. 
> 
> Did you test that booleans and other types display consistently with the current behaviour (like a check mark icon)?

Hey yes it does display the various types consistently. In the example screen it shows a Boolean value as a check mark. I will try to upload additional screenshots and address your comments.

stephenglass commented on plugins/scaffolder-react/src/next/components/ReviewState/ReviewState.tsx:
```diff
@@ -49,30 +82,42 @@ export const ReviewState = (props: ReviewStateProps) => {
                 return [];
               }
+              if (
+                backstageReviewOptions.explode &&
```
Done!

benjdlambert commented:
Does it make sense to do this by default? What's the value behind making this configurable? 😁

stephenglass commented:
> Does it make sense to do this by default? What's the value behind making this configurable? 😁

Giving users flexibility in how the UI looks is a value-add IMO. As for doing this by default, there's definitely a case to be made for that.

benjdlambert commented:
> Giving users flexibility in how the UI looks is a value-add IMO

I'd argue that we already do that by allowing people to provide their own `ReviewStep` component. If there's an issue with us changing the default one, that's the reason that prop is there. I think it makes sense for us to make this change by default, and let's see how it settles? :praey:

stephenglass commented:
@benjdlambert @acierto I've made some notable improvements to this PR.

1. Refactored the schema processing code into a new function that can be recursed to parse deeply nested objects.
2. This refactoring also allowed me to fix an existing Backstage issue where review state logic is not applied to children of objects, like `mask`, `show` options or enum labels. I've added new assertions to the existing unit tests to make sure these review state logic are also being applied to objects.
3. The default behavior of this PR is now to parse nested objects into separate rows by default. To maintain current behavior for an object you can set `ui:backstage.review.explode` option to `false`. This will display the object in a single row rather than breaking out the children into their own row.

Please review when you can and let me know any more feedback! Thanks!!

benjdlambert reviewed (approved):
Nice! Great job! :tada:
